+neuron/Neuron.m

```diff
@@ -0,0 +1,180 @@
+classdef Neuron < dynamicprops
+% Neuron Class for initializing a Neuron session and running generic Neuron functions.
+
+    properties (SetAccess=protected, GetAccess=public)
+        var_list        % List of top-level Neuron variables.
+        fn_double_list  % List of top-level Neuron functions returning a double.
+        fn_string_list  % List of top-level Neuron functions returning a string.
+        object_list     % List of Neuron Objects.
+    end
+
+    methods
+        function self = Neuron()
+        % Initialize the neuron session, if it has not been initialized before.
+        %   Neuron()
+            self = self@dynamicprops;
+            clib.neuron.initialize();
+            arr = split(clib.neuron.get_nrn_functions(), ";");
+            call_list = arr(1:end-1);
+
+            % Add dynamic properties.
+            % TODO: Setting the property crashes for n.L (and perhaps for other variables as well).
+            % Hence, we cannot neatly set all properties upon initialization like we do for neuron.Object.
+            for i=1:length(call_list)
+                f = split(call_list(i), ":");
+                if (f(2) == "263") % variable
+                    self.var_list = [self.var_list f(1)];
+                    p = self.addprop(f(1));
+                    p.SetMethod = self.set_prop(f(1));
+                elseif (f(2) == "280") % function returning a double
+                    self.fn_double_list = [self.fn_double_list f(1)];
+                elseif (f(2) == "296") % function returning a string
+                    self.fn_string_list = [self.fn_string_list f(1)];
+                elseif (f(2) == "325") % object
+                    self.object_list = [self.object_list f(1)];
+                end
+            end
+        end
+        function varargout = subsref(self, S)
+        % If a function is called, but it is not listed as a Neuron class method, try to run it by calling self.call_func_hoc().
+        %   Available functions are displayed using Neuron.list_function().
+        %
+        %   Getting/setting direct top-level variables is possible using:
+        %   n = neuron.Neuron();
+        %   n.t, n.dt, n.GAMMA, n.PHI, etc.
+
+            % S(1).subs is function name;
+            % S(2).subs is a cell array containing arguments.
+            func = S(1).subs;
+
+            % Are we trying to directly access a top-level variable?
+            if (isa(func, "char") && length(S) == 1 && any(strcmp(self.var_list, func)))
+                [varargout{1:nargout}] = clib.neuron.ref(func).get();
+            % Are we trying to directly access a Matlab defined property?
+            elseif (isa(func, "char") && length(S) == 1 && any(strcmp(properties(self), func)))
+                [varargout{1:nargout}] = self.(func);
```

Comment:
We do not yet support further indexing into non-scalar properties? E.g. myobject.vectorproperty(5)?

Reply:
As far as I know this is only relevant for Vector data, for which there is a separate implementation where Vector.data() returns a C++ (clib.array.neuron.Double) array and Vector.double() returns a MATLAB array. Both cannot be indexed directly, e.g. Vector.double(5) does not work.